### PR TITLE
Use commando to access CLN nodes

### PR DESCRIPTION
**Pull request by fiksn**

### CODE REVIEW

File: lightning/api_clnsocket_test.go [hunk 1]
```diff
@@ -291,7 +262,10 @@ func TestClnGetNodeInfoFull(t *testing.T) {
 	})
 	defer closer()
 
-	resp, err := api.GetNodeInfoFull(context.Background(), true, true)
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(15*time.Second))
```
Comment (dcrystalj) [inline lightning/api_clnsocket_test.go:265]: repeating number, move to constant

File: lightning/api_grpc.go [hunk 2]
```diff
@@ -35,13 +34,15 @@ func NewLndGrpcLightningAPI(getData GetDataCall) LightingAPICalls {
...
-	return &LndGrpcLightningAPI{
+	ret := &LndGrpcLightningAPI{
 		Client:       client,
 		RouterClient: routerClient,
 		CleanupFunc:  cleanup,
 		API:          API{GetNodeInfoFullThreshUseDescribeGraph: 500},
-		Name:         "lndgrpc",
 	}
+
+	ret.Name = "lndgrpc"
```
Comment (dcrystalj) [inline lightning/api_grpc.go:44]: why is this change needed ?

File: lightning/api_rest.go [hunk 3]
```diff
@@ -37,13 +37,16 @@ func NewLndRestLightningAPI(getData GetDataCall) LightingAPICalls {
...
-	return &LndRestLightningAPI{
+	ret := &LndRestLightningAPI{
 		Request:   request,
 		Transport: transport,
 		HTTPAPI:   api,
-		Name:      "lndrest",
 		API:       API{GetNodeInfoFullThreshUseDescribeGraph: 500},
 	}
+
+	ret.Name = "lndrest"
```
Comment (dcrystalj) [inline lightning/api_rest.go:47]: why?

File: lightning/connection_commando.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	rune string
+
+	ClnConnection
+}
+
+// Compile time check for the interface
+var _ ClnConnectionAPI = &ClnCommandoConnection{}
+
+// MakeCommandoConnection create a new CLN connection
+func MakeCommandoConnection(addr string, rune string, timeout time.Duration) *ClnCommandoConnection {
```
Comment (dcrystalj) [inline lightning/connection_commando.go:28]: isnt convention in go to use `New` keyword for constructing something

File: lightning/connection_commando.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	ret := &ClnCommandoConnection{}
+
+	ret.addr = addr
+	ret.rune = rune
+	ret.ln = lnsocket.MakeLN(timeout)
+
+	return ret
+}
+
+func convertArgs(param any) string {
```
Comment (dcrystalj) [inline lightning/connection_commando.go:38]: its easier to read if helper methods are below its usage. Also makes sense to se API (public methods) first

File: lightning/connection_commando.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+	err = l.ln.Handshake()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func parseResp(reader io.Reader) (ClnSuccessResp, error) {
```
Comment (dcrystalj) [inline lightning/connection_commando.go:121]: missing test

File: lightning/connection_commando.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	enc := json.NewEncoder(sb)
+	err := enc.Encode(param)
+	if err != nil {
+		return Empty
+	}
+
+	return strings.TrimRight(sb.String(), "\r\n\t")
+}
+
+// Call gets response
```
Comment (dcrystalj) [inline lightning/connection_commando.go:54]: why not naming function GetResponse then ?

File: lightning/connection_common.go [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+// ClnConnection struct
+type ClnConnection struct {
+}
+
+// ClnConnectionAPI interface
+type ClnConnectionAPI interface {
+	// Call calls serviceMethod with args and fills reply with response
+	Call(ctx context.Context, serviceMethod string, args any, reply any) error
+	// StreamResponse is meant for streaming responses it calls serviceMethod with arggs and returns an io.Reader
+	StreamResponse(ctx context.Context, serviceMethod string, args any) (io.Reader, error)
```
Comment (dcrystalj) [inline lightning/connection_common.go:17]: would make sense to unify naming `call` and `streamResponse`. If i understand correctly both are getting response  except one is streaming it.

File: lightning/connection_unix.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+	ret.Client = client
+
+	return ret
+}
+
+// Call helper to call rpc method
+func (l *ClnUnixConnection) Call(ctx context.Context, serviceMethod string, args any, reply any) error {
+	c := make(chan *r.Call, 1)
+
+	go func() { l.Client.Go(serviceMethod, args, reply, c) }()
```
Comment (dcrystalj) [inline lightning/connection_unix.go:44]: why is this wrapped in function. You can call it like this 
```
go l.Client.Go(serviceMethod, args, reply, c)
```

File: lightning/connection_unix.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+	var reply json.RawMessage
+	go func() { l.Client.Go(serviceMethod, args, &reply, c) }()
+
+	select {
+	case call := <-c:
+		if call.Error != nil {
+			return nil, call.Error
+		}
+
+		// TODO: currently we are cheating a bit here by serializing and unserializing
```
Comment (dcrystalj) [inline lightning/connection_unix.go:65]: in comments explaing `WHY` do we need this code and not `what`. I can read the code for `what`

File: lnsocket/commando.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+	return fmt.Errorf("implement commando decode?")
+}
+
+// Encode API
+func (msg *CommandoMsg) Encode(buf *bytes.Buffer, pver uint32) error {
+	if err := lnwire.WriteUint64(buf, msg.RequestID); err != nil {
+		return err
+	}
+
+	buf.WriteString("{\"method\": \"")
```
Comment (dcrystalj) [inline lnsocket/commando.go:58]: I think it would be a bit more readable if you write string with '`' to have multiline ?
```
fmt.Sprintf(`{
  "method": "%s", 
   "params": "%s", 
   "rune":...
`,  msg.Method, msg.Params,...)
```

File: lnsocket/commando.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+	buf.WriteString(msg.Params)
+	buf.WriteString(",\"rune\":\"")
+	buf.WriteString(msg.Rune)
+	buf.WriteString("\"}")
+
+	return nil
+}
+
+// CommandoReadAll reads all commando responses
+func (ln *LN) CommandoReadAll() (string, error) {
```
Comment (dcrystalj) [inline lnsocket/commando.go:70]: test?

File: lightning/api_grpc.go (hunk 2 above)
Comment (fiksn) [reply lightning/api_grpc.go:44]: Property `Name` is used by all implementations and was moved to "base class". And initializing it "inline" looks even more awkward.

File: lightning/api_rest.go (hunk 3 above)
Comment (fiksn) [reply lightning/api_rest.go:47]: Same as in `lndgrpc` case.

File: lightning/connection_common.go (hunk 8 above)
Comment (fiksn) [reply lightning/connection_common.go:17]: For `commando` implementations are a bit different, for `unix socket` `StreamResponse` is just a wrapper around `Call` (but that shouldn't be an issue since reading from local socket should be fast, however you still allocate a lot of memory). I didn't want to add another wrapper and/or make this even less comprehensible.

File: lightning/connection_unix.go (hunk 10 above)
Comment (fiksn) [reply lightning/connection_unix.go:65]: 👍

File: lnsocket/commando.go (hunk 12 above)
Comment (fiksn) [reply lnsocket/commando.go]: This is used by internal lib (will add some tests there)

File: lightning/api_clncommando.go [hunk 13]
```diff
@@ -5,19 +5,19 @@ import (
...
 
 // Compile time check for the interface
-var _ LightingAPICalls = &ClnCommandoAPI{}
+var _ LightingAPICalls = &ClnCommandoLightningAPI{}
 
-// NewClnCommandoLightningAPIRaw gets a new API (not that addr is in pubkey@address format)
+// NewClnCommandoLightningAPIRaw gets a new API (note that addr is in pubkey@address format)
 func NewClnCommandoLightningAPIRaw(addr, rune string) LightingAPICalls {
-	api := &ClnCommandoAPI{}
+	api := &ClnCommandoLightningAPI{}
```
Comment (dcrystalj) [inline lightning/api_clncommando.go:18]: why not return here

File: lnsocket/ln.go [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+		conn, err = brontide.Dial(ln.PrivKeyECDH, netAddr, ln.Timeout, ln.Proxy.Dial)
+	}
+
+	ln.Conn = conn
+
+	return err
+}
+
+// Connect connects to pubkey@host
+func (ln *LN) Connect(s string) error {
```
Comment (dcrystalj) [inline lnsocket/ln.go:224]: Connect function clearly does not do just one thing. Move string parsing in another function

Comment (dcrystalj) [inline lnsocket/ln.go:224]: This can be easily unit tested. Please write tests

File: lnsocket/utils.go [hunk 15]
```diff
@@ -0,0 +1,55 @@
+package lnsocket
```
Comment (dcrystalj) [inline lnsocket/utils.go:1]: missing tests

File: lightning/api_clncommando.go (hunk 13 above)
Comment (fiksn) [reply lightning/api_clncommando.go:18]: Because we initialize some stuff later on, this change was just part of the naming sync.

File: lnsocket/ln.go (hunk 14 above)
Comment (fiksn) [reply lnsocket/ln.go]: 👍 done

**Review by dcrystalj — APPROVED** [review]
